cpp/neetcode_150/07_trees/validate_binary_search_tree.cpp

```diff
@@ -18,57 +9,53 @@
  *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
  * };
  */
+
+// Need a long long int alias because of the constraints of TreeNode.val
+#define ll long long int
+
 class Solution {
 public:
     bool isValidBST(TreeNode* root) {
-        TreeNode* prev = NULL;
-        return inorder(root, prev);
+        return dfs(root, (ll) INT_MIN - 1, (ll) INT_MAX + 1);
     }
+    
 private:
-    bool inorder(TreeNode* root, TreeNode*& prev) {
-        if (root == NULL) {
+    /**
+    * Need long long int versions of standard max and min functions
+    */
+    ll max(ll a, ll b) {
+        return (a > b) ? a : b;
+    }
+    
+    ll min(ll a, ll b) {
+        return (a < b) ? a : b;
+    }
+    
+    /**
+    * DFS through the tree.
+    * 
+    * Algorithm:
+    * 
+    * 1. Return true if root is NULL, since it's a balanced binary search tree.
+    * 2. Checking if root->val is strictly between low and high.
+    * 3. Updating high value if going to left sub-tree.
+    * 4. Updating low value if going through right sub-tree.
+    *
+    * Time Complexity: O(n) [ Visiting every node exactly once ]
+    * Space Complexity: O(n) [ Since in worst case, tree is a linked list, so n function calls in call-stack ]
+    */
+    bool dfs(TreeNode* root, ll low, ll high) {
+                
+        if(root == NULL) {
             return true;
         }
         
-        if (!inorder(root->left, prev)) {
-            return false;
-        }
-        
-        if (prev != NULL && prev->val >= root->val) {
-            return false;
-        }
-        prev = root;
-        
-        if (!inorder(root->right, prev)) {
+        if(root->val > low && root->val < high) {
+            return dfs(root->left, low, min(high, (ll) root->val))
+                && dfs(root->right, max(low, (ll) root->val), high);
+        } else {
```

Comment:
Since you've already verified that root->val is greater than low, and smaller than high, there's no need to calculate ```min(high, root->val)```, and ```max(low, root->val)```.
Recommend changing it to 
```
 return dfs(root->left, low, root->val)
                && dfs(root->right, root->val, high);
```

Reply:
I had indeed overlooked the fact that calculating min and max were unnecessary. Thank you for the suggestion.